rfd/0156-notifications.md

```diff
@@ -0,0 +1,423 @@
+- New features added to Teleport.
+
+## Details
+
+Notifications will be stored in the backend database and be of two types:
+
+1. User-specific notifications: Targets a single specific user. An example is a notification to notify a user that their access request has been approved.
+2. Global notifications: Targets multiple users. An example is a notification to notify users that there is a new access request needing review.
+   They can be configured to target users based on one of the following matchers:
+   1. Permissions: Targets all users with the specified RBAC permission(s).
+   2. Role: Targets all users with the specified role(s).
+   3. All: Targets all users.
+
+Since user-specific notifications are stored by username, notifications for SSO users will persist between sessions.
+
+Notifications will have a default TTL of 30 days, after which they will be automatically deleted. In some cases, notifications may also automatically be deleted prior to their TTL if configured to do so.
+
+#### Backend Database
+
+User-specific notifications will be stored in the database under the key `/notifications/user/<cluster id>/<auth connector>/<the user's username>/<notification id>`.
+
+We include the `cluster id` in the key to prevent the edge case where a user from the root cluster accesses notifications in a leaf cluster that has a user with the same username, this would cause the two users' notifications
+to be merged.
+
+We include the `auth connector` in the key to prevent the edge case where a SSO user has the same username as a local user (or a user using another auth connector), this would cause the users' notifications to be merged.
+
+Global notifications will be stored in the database under the key `/notifications/global/<notification id>`.
+
+There will be a hard cap of 10,000 notifications which can exist at any given time. If a new notification is added beyond this point, it will delete and replace the oldest existing notification.
+
+#### Types and Protos
+
+###### Notification Item
+
+If this is a user-specific notification, it will be stored as-is under `/notification/user/<cluster id>/<auth connector>/<username>/<notification id>`.
+
+If this is a global notification, it will be nested in a [`GlobalNotification`](#global-notifications).
+
+```protobuf
+// Notification represents a notification item.
+message Notification {
+   // id is the UUID of the notification, in UUIDv7 format.
+   string id = 1;
+   // code is the unique notification code.
+   // This code represents the type of notification this is, eg. `access-request-approved`.
+   string code = 2;
+   // title is the title/header of the notification.
+   string title = 3;
+   // description is the text content of the notification.
+   string description = 4;
+   // created is when the notification was created, in UNIX time.
+   google.protobuf.Timestamp created = 5;
+   // ttl is how long the notification should last before automatically expiring. Default is 30 days.
```

Comment:
I would add the unit here. Does a `ttl` of 30 mean 30 days? 30 milliseconds?

Should we use `google.protobuf.Duration` here?